Add Source.lazyInitAsync

swachter opened the pull request:
Allow to create a source lazily and asynchronously. There already is `Sink.lazyInitAsync` and `Flow.lazyInitAsync` but `Source.lazyInitAsync` is missing.

Commit:
add Source.lazyInitAsync

Commit:
add Source.lazyInitAsync

raboof commented:
OK TO TEST

swachter commented:
Can you please give me some information what caused the failure (or help me finding that information). Clicking on the "Jenkins PR Validation - Details" and "continuous-integration/travis-ci/pr - Details" link does not reveal that information.

swachter commented:
The PR Validation fails because of:
```
Caused by: java.io.FileNotFoundException: /home/swachter/projects/cloned/akka/akka-docs/src/main/paradox/stream/operators/Source-or-Flow/lazyInitAsync.md (Datei oder Verzeichnis nicht gefunden)
```
I do not understand where and why that file is referenced. I added the file `Source/lazyInitAsync.md` along the lines of `Flow/lazyInitAsync.md`. Is there some documentation magic that detects that `lazyInitAsync` does now exist both, for flows and for sources? I do not understand how a single lazyInitAsync.md file in "Source-or-Flow" can describe both methods.

raboof commented:
Indeed there is code that identifies when a method is available both on 'Source' and on 'Flow', and in that case expects one documentation page documenting both - since in many cases they have the same meaning, or are related enough that it makes sense to document them in one place. I agree it is slightly awkward in this case, but in general it does seem to make sense.

swachter commented:
Ok - in order to resolve - do I to "merge" the documentation for the two?

raboof commented:
Yes, I think that makes most sense.

Commit:
rename category flow-operators-composed-of-sink-and-sources -> flow-operators; distinguish between Source.lazyInitAsync and Flow.lazyInitAsync documentationwise

swachter commented:
I adjusted the operators documentation slightly:

There are categories for the built-in sinks and sources ("Sink operators" and "Source operators") but there was no general category for built-in flows. I renamed the category "Flow operators composed of Sinks and Sources" into "Flow operators" and added Flow.lazyInitAsync to that category.

In addition, I added some logic to the `StreamOperatorsIndexGenerator` in order to allow to distinguish between certain operators for sources or flows. Now, there are seperate Flow/lazyInitAsync.md and Source/lazyInitAsync.md possible.

Commit:
update generated index

johanandren commented on akka-stream/src/main/scala/akka/stream/impl/LazySourceAsync.scala:
```diff
@@ -0,0 +1,107 @@
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class LazySourceAsync[T, M](sourceFactory: () ⇒ Future[Source[T, M]]) extends GraphStageWithMaterializedValue[SourceShape[T], Future[Option[M]]] {
```
I really think we should avoid another lazy stage here and try to achieve this through composition instead. Tricky with the materialized value though.

johanandren replied:
An alternative is that this should be the _only_ lazy stage and we compose the other factories with it.

johanandren replied:
```Source.lazily(() => Source.fromFutureSource())``` does the trick. We will get the matval semantics of `fromFutureSource` but I think that is better than the sink one as you can easily make your `Future[M]` into a `Future[Option[M]]` with `recover` and `map` but if you try to go the other way around there you will have to invent an exception.

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:
```diff
@@ -437,6 +437,18 @@ object Source {
+   */
+  def lazyInitAsync[T, M](sourceFactory: () ⇒ Future[Source[T, M]]): Source[T, Future[Option[M]]] =
+    Source.fromGraph(new LazySourceAsync[T, M](sourceFactory))
+
```
Unfortunate with the name clash here, since there is a `lazily(() => Source)` one would have wanted it to be `lazilyAsync` which instead would have been better named `lazilySingle` (since there is `Source.single`). 

And the semantics difference where `lazily` fails the returned future if there is no subscription while this completes the future with `None`.

johanandren replied:
We can't change those names though because of binary and source compatibility, the best we can do is to deprecate things.

johanandren commented:
There is a bigger discussion worth to be had here wrt composition ws providing pre-composed operator combinations out of the box.

And then another smaller discussion on aligning the naming and semantics for the async operators.

ping @akka/akka-team

patriknw commented:
Let's take a step back and look at the full picture and try to find some consistent naming. Continue in https://github.com/akka/akka/pull/26704
